### PR TITLE
PT-2563: Pushing a patient with empty data.

### PR DESCRIPTION
**Pull request by veronikaslc**
Fixed.

### CODE REVIEW

**Comment by veronikaslc** [comment]
@AlexLapena @allasm please, test/review

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java [hunk 1]
```diff
@@ -108,8 +108,10 @@ public void save(Patient patient)
                 Integer value = data.get(propertyName);
                 BaseProperty<ObjectPropertyReference> field =
                     (BaseProperty<ObjectPropertyReference>) dataHolder.getField(propertyName);
-                if (field != null && value != null) {
+                if (value != null) {
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java]: what about the "field != null" check?

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java [hunk 2]
```diff
@@ -180,6 +182,9 @@ private boolean hasAnySelected(Collection<String> selectedFieldNames)
                     if (NumberUtils.isDigits(value)) {
                         parsed.put(propertyName, Integer.valueOf(value));
                     }
+                    if (StringUtils.isEmpty(value)) {
+                        parsed.put(propertyName, null);
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java]: do we want a "null" or an empty string here?

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java [hunk 3]
```diff
@@ -507,6 +507,7 @@ private void updateDisordersFromJSON(XWikiDocument doc, BaseObject data, XWikiCo
                 // as in constructor: make unmofidiable
                 this.disorders = Collections.unmodifiableSet(this.disorders);
 
+                data.set(DISORDER_PROPERTIES_OMIMID, null, context);
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: why do we have two data.set(DISORDER_PROPERTIES_OMIMID, ...) on two neighbouring lines?

File: components/patient-data/impl/src/test/java/org/phenotips/data/internal/controller/ParentalAgeControllerTest.java [hunk 4]
```diff
@@ -157,7 +157,7 @@ public void loadMaternalAgeNonZero()
 
         Assert.assertEquals("parentalAge", testData.getName());
         Assert.assertTrue(testData.get(MATERNAL_AGE) == AGE_NON_ZERO);
-        Assert.assertNull(testData.get(PATERNAL_AGE));
+        Assert.assertEquals(AGE_ZERO, testData.get(PATERNAL_AGE));
```
Comment (allasm) [inline components/patient-data/impl/src/test/java/org/phenotips/data/internal/controller/ParentalAgeControllerTest.java]: I believe this test works, but I don't understand why it used to be a null and now it is a 0. Also, what does this case tests, and what does age 0 means - don't we actually want to return null when age is 0?

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/ParentalAgeController.java [hunk 5]
```diff
@@ -181,7 +179,11 @@ public void writeJSON(Patient patient, JSONObject json, Collection<String> selec
         for (String property : getProperties()) {
             if (data.has(property)) {
                 int age = data.getInt(property);
-                result.put(property, age);
+                if (age == 0) {
+                    result.put(property, null);
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/ParentalAgeController.java]: Is there a difference between age 0 and age == null? Don't both logically mean "no age is set", and if yes, isn't "null" a better indication?

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java [hunk 6]
```diff
@@ -114,14 +114,17 @@
     public void writeJSON(Patient patient, JSONObject json, Collection<String> selectedFieldNames)
     {
         PatientData<T> data = patient.getData(getName());
+        String jsonPropertyName = getJsonPropertyName();
+
         if (data == null) {
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java]: it seems we still wont add anything to generated JSON if `data == null` but `selectedFieldNames` has a field this controller controls (as opposed to adding empty values that we should in this case)

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java (hunk 3 above)
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: to overwrite

Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: as discussed, +1

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java (hunk 1 above)
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java]: @marta, is that right that `field` can not be null unless `dataHolder`, which is `doc.getXObject(Patient.CLASS_REFERENCE)` is incorrect and does not have the correct `propertyName` (which is supposedly should, if the version of the code matches the version of the classes in DB)? In which case we probably do want a `NullPointerException` so that we see there is a problem in the installation (class mismatches code)

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java (hunk 2 above)
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/APGARController.java]: as discussed, we probably want a separate issue to consistently represent missing data with empty strings or nulls, and not a random mix of 0-s, null-s and empty strings like we do now

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/ParentalAgeController.java [hunk 7]
```diff
@@ -181,7 +184,11 @@ public void writeJSON(Patient patient, JSONObject json, Collection<String> selec
         for (String property : getProperties()) {
             if (data.has(property)) {
                 int age = data.getInt(property);
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/ParentalAgeController.java]: we should always use optInt(property,0) to correctly handle misformatted data (will also handle empty strings ans nulls instead of 0s as indicators for the missing age)

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java [hunk 8]
```diff
@@ -491,7 +494,10 @@ private JSONArray joinArrays(JSONArray jsonOne, JSONArray jsonTwo)
     private void updateDisordersFromJSON(XWikiDocument doc, BaseObject data, XWikiContext context, JSONObject json)
     {
         try {
-            JSONArray inputDisorders = json.optJSONArray(JSON_KEY_DISORDERS);
+            if (!json.has(JSON_KEY_DISORDERS)) {
+                return;
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: why this change? optJSONArray(JSON_KEY_DISORDERS) would return null if either there is no key or key value is not an array (e.g. misformatted JSON)

with new code, no key is covered, by value not being a proper JSON array will result in a JSONException

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java [hunk 9]
```diff
@@ -202,7 +202,7 @@ private Object format(String key, Object value)
     /** For converting JSON into internal representation. */
     private Object inverseFormat(String key, Object value)
     {
-        if (value != null) {
+        if (value != null && !value.equals(null)) {
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java]: just curious how does value.equals(null) work, is it for JSONObject.NULL?

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java [hunk 10]
```diff
@@ -290,23 +306,25 @@ public void save(Patient patient)
             XWikiDocument doc = (XWikiDocument) this.documentAccessBridge.getDocument(patient.getDocument());
             BaseObject dataHolder = doc.getXObject(getXClassReference());
             PatientData<T> data = patient.getData(this.getName());
-            if (dataHolder == null && data != null) {
+            if (dataHolder == null || dataHolder == null) {
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java]: dataHolder == null || dataHolder == null ?

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AllergiesController.java [hunk 11]
```diff
@@ -175,15 +175,15 @@ public void writeJSON(Patient patient, JSONObject json, Collection<String> selec
     public PatientData<String> readJSON(JSONObject json)
     {
         List<String> result = new ArrayList<>();
+        if (!json.has(DATA_NAME)) {
+            return null;
+        }
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AllergiesController.java]: minor: move this before you create result = new ArrayList()

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java (hunk 10 above)
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java]: I think with that change - https://github.com/phenotips/phenotips/pull/2047/commits/9aa4f6ae0423e0765ea08eb3988df7de22b5d30f#diff-a359c7d8e774fa88f6b015042ad1f9a0R179 - data will actually be null, and the old code was correct?

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java (hunk 9 above)
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java]: yup

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java (hunk 10 above)
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java]: had to be data, typo

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java (hunk 8 above)
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: actually yes, didn't think about not array case

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/ParentalAgeController.java (hunk 7 above)
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/ParentalAgeController.java]: bump

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java [hunk 12]
```diff
@@ -287,26 +306,25 @@ public void save(Patient patient)
...
-                    continue;
-                }
-                if (this.getCodeFields().contains(propertyName) && this.isCodeFieldsOnly()) {
-                    @SuppressWarnings("unchecked")
-                    List<VocabularyProperty> terms = (List<VocabularyProperty>) propertyValue;
-                    List<String> listToStore = new LinkedList<>();
-                    for (VocabularyProperty term : terms) {
-                        String name = StringUtils.isNotBlank(term.getId()) ? term.getId() : term.getName();
-                        listToStore.add(name);
+                if (isInKeySet(data, propertyName)) {
```
Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/AbstractComplexController.java]: after PT-2670 is merged in, we can just write if (data.hasKey(propertyName)) .... and remove isInKeySet()

**Comment by sdumitriu** [comment]
Manually merged.
